### PR TITLE
5-arg mul! bug fixes

### PR DESCRIPTION
**Pull request by tkf**
Another followup of #29634:

* Enforce strong-zero behavior for alpha
* Fix 5-arg mul! for Bi/Tri/Sym * Diag

### CODE REVIEW

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 1]
```diff
@@ -282,76 +282,78 @@ end
...
+    out .= adjoint.(A.parent.diag) .* in .* _strong0(alpha) .+ _scaledout(out, beta)
 @inline mul!(out::AbstractMatrix, A::Transpose{<:Any,<:Diagonal}, in::StridedMatrix,
              alpha::Number, beta::Number) =
-    out .= transpose.(A.parent.diag) .* in .* alpha .+ _scaledout(out, beta)
+    out .= transpose.(A.parent.diag) .* in .* _strong0(alpha) .+ _scaledout(out, beta)
 
 @inline mul!(out::AbstractMatrix, A::Diagonal, in::Adjoint{<:Any,<:StridedMatrix},
              alpha::Number, beta::Number) =
-    out .= A.diag .* in .* alpha
+    out .= A.diag .* in .* _strong0(alpha) .+ _scaledout(out, beta)
```
Comment (tkf) [inline stdlib/LinearAlgebra/src/diagonal.jl]: beta was not used (and the test was not catching this!)

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 2]
```diff
@@ -382,11 +384,11 @@ mul!(C::AbstractMatrix, A::Transpose{<:Any,<:Diagonal}, B::Transpose{<:Any,<:Rea
 @inline mul!(C::AbstractMatrix,
              A::Adjoint{<:Any,<:Diagonal}, B::Adjoint{<:Any,<:RealHermSymComplexSym},
              alpha::Number, beta::Number) =
-    C .= adjoint.(A.parent.diag) .* B .* alpha .+ C .* beta
+    C .= adjoint.(A.parent.diag) .* B .* _strong0(alpha) .+ _scaledout(C, beta)
```
Comment (tkf) [inline stdlib/LinearAlgebra/src/diagonal.jl]: weak beta was used and the test was not catching this

File: stdlib/LinearAlgebra/src/diagonal.jl (hunk 1 above)
Comment (tkf) [reply stdlib/LinearAlgebra/src/diagonal.jl]: tests added by 1b961615340292b7416e94a272b11787dea3d617

File: stdlib/LinearAlgebra/src/diagonal.jl (hunk 2 above)
Comment (tkf) [reply stdlib/LinearAlgebra/src/diagonal.jl]: tests added in f71905fac262fda255de927086047efcbf45f834

**Comment by tkf** [comment]
I ran the "full test" on 9da77a493f05b7ba8c58620d9f93c0aebcc09957 by commenting out this line

https://github.com/JuliaLang/julia/blob/9da77a493f05b7ba8c58620d9f93c0aebcc09957/stdlib/LinearAlgebra/test/addmul.jl#L121

All tests passed (after more than 6 hours...).  As a side note, I find `--compile=min` helps a lot as the bottleneck of this test is compilation.

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 3]
```diff
@@ -277,81 +277,83 @@ function rmul!(A::AbstractMatrix, transB::Transpose{<:Any,<:Diagonal})
 end
 
 # Elements of `out` may not be defined (e.g., for `BigFloat`). To make
-# `mul!(out, A, B)` work for such cases, `_scaledout` short-circuits
+# `mul!(out, A, B)` work for such cases, `out .*ₛ beta` short-circuits
```
Comment (andreasnoack) [inline stdlib/LinearAlgebra/src/diagonal.jl]: This doesn't render for me. Which character are you using here?

**Review by andreasnoack — COMMENTED** [review]
LGTM except for the rending issue

File: stdlib/LinearAlgebra/src/diagonal.jl (hunk 3 above)
Comment (tkf) [reply stdlib/LinearAlgebra/src/diagonal.jl]: It's `*_s` (for short-circuiting):

image

Maybe `*ˢ` (superscript `s`) or `*′` (prime)?

Or we can pick a non-standard character (maybe `⋇`?) from:

`* / ÷ % & ⋅ ∘ × |\\| ∩ ∧ ⊗ ⊘ ⊙ ⊚ ⊛ ⊠ ⊡ ⊓ ∗ ∙ ∤ ⅋ ≀ ⊼ ⋄ ⋆ ⋇ ⋉ ⋊ ⋋ ⋌ ⋏ ⋒ ⟑ ⦸ ⦼ ⦾ ⦿ ⧶ ⧷ ⨇ ⨰ ⨱ ⨲ ⨳ ⨴ ⨵ ⨶ ⨷ ⨸ ⨻ ⨼ ⨽ ⩀ ⩃ ⩄ ⩋ ⩍ ⩎ ⩑ ⩓ ⩕ ⩘ ⩚ ⩜ ⩞ ⩟ ⩠ ⫛ ⊍ ▷ ⨝ ⟕ ⟖ ⟗`

Comment (tkf) [reply stdlib/LinearAlgebra/src/diagonal.jl]: We can also use normal function like `mulsc`.  The code would look like:

    mulsc.(A.diag .* in, alpha) .+ mulsc.(out, beta)

Comment (JeffBezanson) [reply stdlib/LinearAlgebra/src/diagonal.jl]: According to wikipedia ⊙ is sometimes used for element-wise product.

Comment (staticfloat) [reply stdlib/LinearAlgebra/src/diagonal.jl]: `⊙` is indeed the notation that my profs used for element-wise product.

**Comment by JeffBezanson** [comment]
We can worry about the choice of operator later, to get these fixes in.

File: stdlib/LinearAlgebra/src/diagonal.jl (hunk 3 above)
Comment (tkf) [reply stdlib/LinearAlgebra/src/diagonal.jl]: We need a non-standard binary operator and directly define the broadcasted behavior of it.  (The usecase here is to hijack the broadcasted behavior to implement short-circuiting at the level of vector-vector multiplication.)

So using `⊙` would mean to write `.⊙` which looks a bit strange as it sounds like doing "doubly" element-wise operation on nested objects.

Comment (timholy) [reply stdlib/LinearAlgebra/src/diagonal.jl:280]: [TensorCore.jl](https://github.com/JuliaMath/TensorCore.jl)

**Comment by timholy** [comment]
On one of my machines, the `LinearAlgebra/addmul` tests take 3 times as long as the third-slowest test and twice as long as the second-slowest test. That means if you're running Julia's tests locally you spend most of the time waiting for this one test to finish. Is this something we should consider optimizing, or given buildkite is it not worth it?

Of course, it might also be a good compiler benchmark, maybe worth investigating.

**Comment by timholy** [comment]
Part of the answer:

```julia
julia> m = only(methods(LinearAlgebra._generic_matmatmul!))
_generic_matmatmul!(C::AbstractVecOrMat{R}, tA, tB, A::AbstractVecOrMat{T}, B::AbstractVecOrMat{S}, _add::LinearAlgebra.MulAddMul) where {T, S, R} in LinearAlgebra at /home/tim/src/julia-branch2/usr/share/julia/stdlib/v1.9/LinearAlgebra/src/matmul.jl:850

julia> length(filter(mi -> mi !== nothing, collect(m.specializations)))
5272
```
:laughing:

**Pull request merged**